0000-benchmark-suite.md

```diff
@@ -0,0 +1,782 @@
+shuffle the order of object files, shuffle object files within `.rlib`s (which
+are `.a` files that also contain a metadata file), and finally invoke the
+underlying linker. Before running benchmarks, we will create a few different
+`wasmtime` binaries that we choose from when sampling different benchmark
+executions, each with a different, randomized object file link order. This will
+help us avoid measurement bias related to getting "lucky" or "unlucky" with a
+single code layout due to the order of object files passed to the linker.
+
+> Note that this approach is more coarsely grained than what Stabilizer
+> achieves. Stabilizer randomized code layout at the function granularity, while
+> this approach works at the object file granularity.
+
+### Randomizing Benchmark Execution Order
+
+If we are running benchmarks with two different versions of Wasmtime, version A
+and version B, we should not take all samples from version A and then all
+samples from version B. We should, instead, intersperse executions of version A
+and version B. While we should always ensure that the CPU frequency scaling
+governor is set to "performance" when running benchmarks, this additionally
+helps us avoid some measurement bias from CPU state transitions that aren't
+constrained within the duration of process execution, like dynamic CPU
+throttling due to overheating.
+
+## Analysis
+
+The worst kind of analysis we can do is something like compare averages with and
+without our proposed optimization, and use that to conclude whether the
+optimization "works". Or, equivalently, eyeball two bars in a graph and decide
+which one looks better. We don't know whether the difference is meaningful or
+just noise.
+
+A better approach is to use a test of statistical significance to answer the
+question as to whether the difference is meaningful or not. For comparing two
+means we would use [Student's
+t-test](https://en.wikipedia.org/wiki/Student%27s_t-test) and for comparing
+three or more means we would use [analysis of variance
+(ANOVA)](https://en.wikipedia.org/wiki/Analysis_of_variance). These tests allow
+us to justify statements like "we are 95% confident that change X produces a
+meaningful difference in execution time".
+
+We aren't, however, solely interested in whether execution time is different or
+not. We want to know how much of a speed up an optimization yields! A
+significance test, while an improvement over eyeballing results, is still
+"asking the wrong question". This leads us to the best approach for analyzing
+performance results: [effect size confidence
+intervals](https://en.wikipedia.org/wiki/Effect_size). This analysis justifies
+statements like "we are 95% confident that change X produces a 5.5% (+/- 0.8%)
+speed up compared to the original system". See the "Challenge of Summarizing
+Results" and "Measuring Speedup" sections of *Rigorous Benchmarking in
+Reasonable Time*<sup>[4](#ref-4)</sup> for further details.
```

Comment:
Should there be more bench runs when the difference on a benchmark is less statistically significant?

Reply:
The referenced paper grapples with this question, and ultimately gives you a formula where you can plug in the levels of variation (taken from some initial experiments) at each layer and a desired confidence interval and then it tells you how many iterations at each level you need to perform in your experiment before you actually conduct it. It's a good paper, I recommend reading it!

It is *not* recommended to at runtime keep running the benchmark for X minutes or until a given confidence interval is reached because this makes results between different runs (e.g. with and without your changes) subtly incomparable.